Update NCCL to 2.23.4

kwen2501 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142179

Major features:
(2.22)
- Lazy data channel connection to speed up init and lower memory usage when not using all algorithms.
- Cost estimation API.
- Support ncclGroupStart/End for ncclCommAbort/Destroy.

(2.23)
- Scalable initialization API.
- PAT (Parallel Aggregated Trees) algorithm for allgather and reduce scatter operations using one GPU per node.
- Accelerate intra-node communication when buffers are registered.

Since `nvidia-nccl-cu11` does not have version 2.23, we are keeping 2.21 for CUDA 11.8 build.

kwen2501 committed:
Update NCCL to 2.23.4

[ghstack-poisoned]

atalman commented on .github/scripts/generate_binary_build_matrix.py:
```diff
@@ -44,7 +44,7 @@
         "nvidia-curand-cu11==10.3.0.86; platform_system == 'Linux' and platform_machine == 'x86_64' | "
         "nvidia-cusolver-cu11==11.4.1.48; platform_system == 'Linux' and platform_machine == 'x86_64' | "
         "nvidia-cusparse-cu11==11.7.5.86; platform_system == 'Linux' and platform_machine == 'x86_64' | "
-        "nvidia-nccl-cu11==2.21.5; platform_system == 'Linux' and platform_machine == 'x86_64' | "
```
Please keep cuda 11.8 builds as is without the change.

atalman replied:
See https://github.com/pytorch/pytorch/issues/138609

kwen2501 replied:
You are right. Thanks for catching it. Fixed.

kwen2501 committed:
Update on "Update NCCL to 2.23.4"


Major features:
(2.22)
- Lazy data channel connection to speed up init and lower memory usage when not using all algorithms.
- Cost estimation API.
- Support ncclGroupStart/End for ncclCommAbort/Destroy.

(2.23)
- Scalable initialization API.
- PAT (Parallel Aggregated Trees) algorithm for allgather and reduce scatter operations using one GPU per node.
- Accelerate intra-node communication when buffers are registered.



[ghstack-poisoned]

atalman commented on .ci/docker/common/install_cuda.sh:
```diff
@@ -2,7 +2,7 @@
 set -ex
 
-NCCL_VERSION=v2.21.5-1
+NCCL_VERSION=v2.23.4-1
```
hi @kwen2501 I believe here need to pin ``NCCL_VERSION_CU118=v2.21.5-1``
and use it in ``function install_118``

kwen2501 replied:
I was just wondering about it :) Thanks!
I followed the `CUDNN_VERSION` practice and put an overwrite in `function install_118` now.

kwen2501 committed:
Update on "Update NCCL to 2.23.4"


Major features:
(2.22)
- Lazy data channel connection to speed up init and lower memory usage when not using all algorithms.
- Cost estimation API.
- Support ncclGroupStart/End for ncclCommAbort/Destroy.

(2.23)
- Scalable initialization API.
- PAT (Parallel Aggregated Trees) algorithm for allgather and reduce scatter operations using one GPU per node.
- Accelerate intra-node communication when buffers are registered.



[ghstack-poisoned]

kwen2501 commented:
Hi @atalman for this CI failure
> [Lint / workflow-checks / linux-job](https://hud.pytorch.org/pr/pytorch/pytorch/142179#34012547282) ([gh](https://github.com/pytorch/pytorch/actions/runs/12192235537/job/34012547282))
RuntimeError: NCCL submodule version 2.23.4 differs from wheel version 2.21.5

Do you know if this is just a "superficial" check, or should we actually checkout the commit of 2.21.5 when building with CUDA 11.8?

Skylion007 commented:
> Hi @atalman for this CI failure
> 
> > [Lint / workflow-checks / linux-job](https://hud.pytorch.org/pr/pytorch/pytorch/142179#34012547282) ([gh](https://github.com/pytorch/pytorch/actions/runs/12192235537/job/34012547282))
> > RuntimeError: NCCL submodule version 2.23.4 differs from wheel version 2.21.5
> 
> Do you know if this is just a "superficial" check, or should we actually checkout the commit of 2.21.5 when building with CUDA 11.8?

See: https://github.com/pytorch/pytorch/issues/138609 for more discussion, this will cause static and dynamically linked builds of PyTorch to diverge, unfortunately. I think this mostly affected conda builds previously (and maybe nightly ones too), but we are dropping the official PT conda builds. @malfet Added this check, I think.

malfet commented:
@kwen2501 do you need it for PyTorch-2.6?

Imo what we should do (I remember talking about it 3+ month ago) are to build with nccl from source for 11.8 and use pypi for other builds

kwen2501 commented:
@malfet no, I synced with @atalman, no critical issue that makes 2.23 a must for 2.6.
The plan is to land the upgrade after the branch cut.
> are to build with nccl from source for 11.8 and use pypi for other builds

Yeah, @atalman and I agree with this too. For the nightlies (after 2.6) of 11.8, we will have static build (large wheel) built from source code of NCCL 2.23, so that we avoid a divergence of commit point. For CUDA 12 builds, we will use dynamic linking.

The risk of NCCL 2.23 + CUDA 11.8 should be controllable because (1) it is only for nightlies; and (2) if people found issues, they will report it. And by the time of 2.7, we may not need 11.8 build anymore.

main-horse commented:
Hi, one note of caution for this pull request.

NCCL 2.23.4 disables DirectP2P for certain PCIe >2GPU nodes. See the relevant NCCL [commit](https://github.com/NVIDIA/nccl/commit/68b542363f9a44cdaac480f51ebe0fc26de96139#diff-2d7668f047805397593254bbc783171e216bbed8dfb83936876358491b94f003L280):

![image](https://github.com/user-attachments/assets/2cebbada-d222-4608-a457-b8a47e81f1d5)

This causes **drastic performance drops** in all distributed collectives for many-GPU PCIe nodes. On [nccl-tests](https://github.com/nvidia/nccl-tests) `all_reduce_perf`, I see a **3x reduction** in busbw from `7.5GB/s` -> `2.5GB/s` on my 8x3090 node.

Although I have not empirically verified this, I believe the above slowdown will apply for *all* PCIe GPU nodes that happen to have amd CPUs, including nodes with datacenter PCIe NVIDIA GPUs.

---

On the end user side, this can be easily remedied by forcing `NCCL_P2P_LEVEL=SYS`, which restores original performance.

However, even though that solution is simple to implement, I think it is reasonable to anticipate many users complaining about 'reduced training performance' if the environment variable is not enabled out-of-the-box.

What are your thoughts?

aws-nslick commented:
[nccl 2.24.3](https://github.com/NVIDIA/nccl/commit/6aae37927840b2bd7b7d42d2f0050e75f88ee97f) is now released and might be worth targeting instead of [nccl 2.23.4](https://github.com/NVIDIA/nccl/commit/68b542363f9a44cdaac480f51ebe0fc26de96139). Full diff: [link](https://github.com/NVIDIA/nccl/compare/68b542363f9a44cdaac480f51ebe0fc26de96139...6aae37927840b2bd7b7d42d2f0050e75f88ee97f).

kwen2501 commented:
@main-horse thanks for the heads-up!

> Although I have not empirically verified this, I believe the above slowdown will apply for all PCIe GPU nodes that happen to have amd CPUs, including nodes with datacenter PCIe NVIDIA GPUs.

Per the diff you showed, the P2P change does not seem to apply to systems with AMD CPUs -- they still have `p2plevel = PATH_SYS`. 

@sjeaugey @KaimingOuyang @AddyLaddy @jbachan wondering if you have any comments re the 3090+PCIe perf change?

kwen2501 commented:
> [nccl 2.24.3](https://github.com/NVIDIA/nccl/commit/6aae37927840b2bd7b7d42d2f0050e75f88ee97f) is now released and might be worth targeting instead of [nccl 2.23.4](https://github.com/NVIDIA/nccl/commit/68b542363f9a44cdaac480f51ebe0fc26de96139). Full diff: [link](https://github.com/NVIDIA/nccl/compare/68b542363f9a44cdaac480f51ebe0fc26de96139...6aae37927840b2bd7b7d42d2f0050e75f88ee97f).

Yeah it seems NCCL 2.24 has nice features and fixes. On the other hand, it seems to be further ahead of CUDA 11.8 which our nightly build still needs to support? 

@atalman @malfet do we have a build plan decided for pytorch 2.7? Are we dropping CUDA 11.8 from both the release and the nightly?

main-horse commented:
> does not seem to apply to systems with <=2 AMD CPUs

@kwen2501 the compar is `system->nodes[GPU].count <= 2`, which means less-than-or-equal 2 **GPUs**. To avoid all doubt: it triggers on both single and dual socket AMD CPU systems, and will not trigger at all on Intel CPU systems.

kwen2501 commented:
You are absolutely right, @main-horse. I misread the code. Thanks for the correction. 

In that regards, I also found NCCL's rationale for disabling the P2P:
(from 2.23's commit message)
> Set P2P level to PXB on AMD CPUs when using more than 2 GPUs per node
> * P2P would cause a significant performance degradation when using
>   many GPUs, and therefore many interleaved data flows.
> * Disable P2P through the CPU when we have 3+ GPUs per node; keep it
>   enabled when we only have 2 GPUs.

@sjeaugey it seems the code change would disable P2P for *all* CPU types first, with an exception for AMD + 2 GPUs. That seems slightly broader than the commit message above. Just want to confirm if that's intended.

aws-nslick commented:
> Although I have not empirically verified this, I believe the above slowdown will apply for all PCIe GPU nodes that happen to have amd CPUs, including nodes with datacenter PCIe NVIDIA GPUs.

For the datacenter case I think the assumption is that nvlink will (usually) be available, and the later `path->type <= p2pLevel` will always be true (`path->type` should always be 1 between any two gpus, `p2pLevel` will assume either 4 or 7, defined [here](https://github.com/NVIDIA/nccl/blob/master/src/graph/topo.h#L42-L82))

aws-nslick commented:
posting for context: https://github.com/pytorch/pytorch/pull/146073 / https://github.com/pytorch/pytorch/commit/eecee5863e698d19458b33df7bfecbda0a04557a merged today and bumped HEAD to NCCL 2.25 across all 12.x toolkits, so this can very likely be pruned.
